etc/migration/migrateUsersToShibboleth.js

```diff
@@ -0,0 +1,388 @@
+ *
+ * @param  {Array}      allUsers                The array of all the users for a tenancy
+ * @param  {Function}   callback                Standard callback function
+ * @param  {Object}     callback.err            An error that occurred, if any
+ * @param  {Array}      callback.mappedUsers    The Shibboleth users that were mapped to existing users
+ * @api private
+ */
+var _mapUsersToShibUsers = function(allUsers, callback) {
+    var shibbolethUsers = _.filter(allUsers, function(user) {
+        return user.loginId && user.loginId.split(':')[1] === 'shibboleth';
+    });
+
+    if (shibbolethUsers.length < 1) {
+        log().info('No Shibboleth users found for tenant %s', tenantAlias);
+        _exit(0);
+    }
+
+    log().info('Found %s users with Shibboleth accounts', shibbolethUsers.length);
+
+    var nonShibUsers = _.difference(allUsers, shibbolethUsers);
+
+    var mappedUsers = [];
+
+    _.each(shibbolethUsers, function(userHash) {
+        log().info('Processing user %s', userHash.email);
+
+        // Find Shibboleth EPPN for each user (if exists)
+        _findShibbolethEppn(userHash.loginId, function(err, eppn) {
+            if (!eppn) {
+                _writeErrorRow(userHash, null, 'This Shibboleth user has no EPPN');
+            }
+
+            // Extend the user object with the EPPN
+            var userWithEppn = _.extend(userHash, {'eppn': eppn});
+
+            // Find a non-Shibboleth user account with an email that matches the EPPN
+            _getUserMatches(nonShibUsers, userWithEppn, function(err, userMatch) {
+                if (err) {
+                    log().warn(err);
+                }
+
+                // Link the Shibboleth account to the old user account
+                _swapUserLogin(userMatch, userWithEppn, function(err) {
+                    if (err) {
+                        _writeErrorRow(userWithEppn, userMatch, 'Failed to link Shibboleth login to existing user');
+                    } else {
+                        mappedUsers.push(userMatch);
+                    }
+
+                    if ((shibbolethUsers.length -1) === shibbolethUsers.indexOf(userHash)) {
+                        _.each(_.difference(nonShibUsers, mappedUsers), function(user) {
+                            _writeErrorRow(null, user, util.format('No matching Shibboleth user found for user %s', user.email));
+                        });
+
+                        callback(null, mappedUsers);
+                    }
+                });
+            });
+        });
+    });
+};
+
+/**
+ * Write errors to a CSV file
+ *
+ * @param  {Object}     shibUser                The Shibboleth user involved
+ * @param  {Object}     oldUser                 The existing user we want to map to
+ * @param  {String}     message                 A short message detailing the issue
+ * @api private
+ */
+var _writeErrorRow = function(shibUser, oldUser, message) {
```

Comment:
This can be changed to `function _writeErrorRow`. There is a subtle yet important difference between `var _writeErrorRow = function(...` and `function _writeErrorRow(...`.

* When using `var`, the `_writeErrorRow` symbol will be available anywhere in the scope that it's declared, however its value will be `undefined` until this area of the code is evaluated. When doing lots of asynchronous operations, that typically isn't an issue, however if you try and invoke `_writeErrorRow` at the top of this file before you hit an async function, you'll see you have an undefined variable.
* When doing `function`, you'll find that the symbol is available at the beginning of the file *and* it's definition is set to the function, so running `_writeErrorRow(myUser, oldUser, 'oops!')` at the top of this file will work as you'd expect